pkg/ring/ring.go

```diff
@@ -75,26 +75,38 @@ type ReadRing interface {
 	HasInstance(instanceID string) bool
 }
 
-// Operation can be Read or Write
-type Operation int
-
-// Values for Operation
-const (
-	Read Operation = iota
-	Write
-	Reporting // Special value for inquiring about health
-
-	// BlocksSync is the operation run by the store-gateway to sync blocks.
-	BlocksSync
-
-	// BlocksRead is the operation run by the querier to query blocks via the store-gateway.
-	BlocksRead
-
-	// Ruler is the operation used for distributing rule groups between rulers.
-	Ruler
+// Operation describes which instances can be included in the replica set, based on their state.
+type Operation interface {
+	// ShouldExtendReplicaSet returns true if given a state of instance that's going to be
+	// added to the replica set, the replica set size should be extended by 1
+	// more instance for the given operation.
+	ShouldExtendReplicaSetOnState(s IngesterState) bool
+
+	// Used during "filtering" phase to remove undesired instances based on their state.
+	IncludeInstanceInState(s IngesterState) bool
+}
 
-	// Compactor is the operation used for distributing tenants/blocks across compactors.
-	Compactor
+var (
+	// Write operation that also extends replica set, if ingester state is not ACTIVE.
+	Write Operation = NewOpWithReplicaSetExtension(func(s IngesterState) bool {
+		// We do not want to Write to Ingesters that are not ACTIVE, but we do want
+		// to write the extra replica somewhere.  So we increase the size of the set
+		// of replicas for the key. This means we have to also increase the
```

Comment:
The comment about read has been copy-pastes but doesn't apply anymore to this piece of code, which is Write specific.